src/mongo/db/pipeline/expression_test.cpp

```diff
@@ -2208,6 +2208,132 @@ TEST(ExpressionPowTest, NegativeOneRaisedToNegativeOddExponentShouldOutPutNegati
+        expCtx, fromjson("{ $indexOfArray : [ undefined , 1, 1]}"), expCtx->variablesParseState);
+
+    auto isExpIndexOfArray = dynamic_cast<ExpressionIndexOfArray*>(expIndex.get());
+    ASSERT_TRUE(isExpIndexOfArray);
+
+    auto nullishValueOptimizedToExpConstant = isExpIndexOfArray->optimize();
+    auto shouldBeExpressionConstant = dynamic_cast<ExpressionConstant*>(nullishValueOptimizedToExpConstant.get());
+    ASSERT_TRUE(shouldBeExpressionConstant);
+    // Nullish input array should become a Value(BSONNULL).
+    ASSERT_VALUE_EQ(Value(BSONNULL), shouldBeExpressionConstant->getValue());
+}
+
+TEST(ExpressionIndexOfArray,
+     OptimizedExpressionIndexOfArrayWithConstantArgumentsShouldEvaluateProperly) {
+
+    intrusive_ptr<ExpressionContextForTest> expCtx(new ExpressionContextForTest());
+
+    auto expIndexOfArray =
+        Expression::parseExpression(expCtx,
+                                    // Search for 2.
+                                    fromjson("{ $indexOfArray : [ [0, 1, 2, 3, 4, 5] , 2] }"),
+                                    expCtx->variablesParseState);
+    auto optimizedIndexOfArray = expIndexOfArray->optimize();
+    ASSERT_VALUE_EQ(Value(2), optimizedIndexOfArray->evaluate(Document()));
+
+    auto IndexNotFound =
+        Expression::parseExpression(expCtx,
+                                    // Search for 10.
+                                    fromjson("{ $indexOfArray : [ [0, 1, 2, 3, 4, 5] , 10] }"),
+                                    expCtx->variablesParseState);
+    auto optimizedIndexNotFound = IndexNotFound->optimize();
+    // Should evaluate to -1 if not found.
+    ASSERT_VALUE_EQ(Value(-1), optimizedIndexNotFound->evaluate(Document()));
+}
+
+TEST(ExpressionIndexOfArray,
+     OptimizedExpressionIndexOfArrayWithConstantArgumentsShouldEvaluateProperlyWithRange) {
+    intrusive_ptr<ExpressionContextForTest> expCtx(new ExpressionContextForTest());
+
+    auto expIndexOfArray =
+        Expression::parseExpression(expCtx,
+                                    // Search for 4 between 3(inclusive) and 5 (exclusive).
+                                    fromjson("{ $indexOfArray : [ [0, 1, 2, 3, 4, 5] , 4, 3, 5] }"),
+                                    expCtx->variablesParseState);
+    auto optimizedIndexOfArray = expIndexOfArray->optimize();
+    ASSERT_VALUE_EQ(Value(4), optimizedIndexOfArray->evaluate(Document()));
+
+    auto IndexNotFoundInRange =
+        Expression::parseExpression(expCtx,
+                                    // Search for 0 between 3(inclusive) and 5 (exclusive).
+                                    fromjson("{ $indexOfArray : [ [0, 1, 2, 3, 4, 5] , 0, 3, 5] }"),
+                                    expCtx->variablesParseState);
+    auto optimizedIndexNotFoundInRange = IndexNotFoundInRange->optimize();
+    // Should evaluate to -1 if not found in range.
+    ASSERT_VALUE_EQ(Value(-1), optimizedIndexNotFoundInRange->evaluate(Document()));
+}
+
```

Comment:
Can you add a test where there are duplicate values in the array? Such a test should have caught the bug you mentioned with inserting into the unordered_map twice.